async_simple/util/ThreadPool.h

```diff
@@ -25,372 +25,137 @@
+    explicit ThreadPool(size_t threadNum = DEFAULT_THREAD_NUM,
+                        bool enableWorkSteal = false);
     ~ThreadPool();
 
-private:
-    ThreadPool(const ThreadPool &);
-    ThreadPool &operator=(const ThreadPool &);
-
-public:
-    ERROR_TYPE scheduleById(std::function<void()> fn, int32_t id = -1);
-    int32_t getCurrentId() const;
+    ThreadPool::ERROR_TYPE scheduleById(std::function<void()> fn,
+                                        size_t id = -1);
+    size_t getCurrentId() const;
     size_t getItemCount() const;
     size_t getThreadNum() const { return _threadNum; }
-    bool start();
-    void stop(STOP_TYPE stopType = STOP_AFTER_QUEUE_EMPTY);
-    void clearQueue();
-    size_t getActiveThreadNum() const;
-    const std::vector<std::thread> &getThreads() const { return _threads; }
-
-private:
-    void push(WorkItem *item);
-    void pushFront(WorkItem *item);
-    bool tryPopItem(WorkItem *&);
-    bool createThreads();
-    void workerLoop(size_t id);
-    void waitQueueEmpty();
-    void stopThread();
-    std::pair<int32_t, ThreadPool *> *getCurrent() const;
 
 private:
+    std::pair<size_t, ThreadPool *> *getCurrent() const;
     size_t _threadNum;
-    std::vector<std::queue<WorkItem *> > _queue;
+    std::vector<Queue<WorkItem>> _queues;
     std::vector<std::thread> _threads;
-    mutable std::vector<ProducerConsumerCond> _cond;
-    volatile bool _push;
-    volatile bool _run;
-    std::atomic<std::uint32_t> _activeThreadCount;
-
-private:
-    friend class ThreadPoolTest;
+    bool _enableWorkSteal;
 };
 
-typedef std::shared_ptr<ThreadPool> ThreadPoolPtr;
+inline ThreadPool::ThreadPool(size_t threadNum, bool enableWorkSteal)
+    : _threadNum(threadNum ? threadNum : DEFAULT_THREAD_NUM),
+      _queues(_threadNum),
+      _enableWorkSteal(enableWorkSteal) {
+    auto worker = [this](size_t id) {
+        auto current = getCurrent();
+        current->first = id;
+        current->second = this;
+        while (true) {
+            WorkItem workerItem = {};
+            if (_enableWorkSteal) {
+                // Try to do work steal firstly.
+                for (auto n = 0; n < _threadNum * 2; ++n) {
+                    if (_queues[(id + n) % _threadNum].try_pop(
+                            workerItem,
+                            [](auto &item) { return item.autoSchedule; }))
+                        break;
+                }
+            }
 
-inline ThreadPool::ThreadPool(size_t threadNum)
-    : _threadNum(threadNum),
-      _queue(threadNum ? threadNum : DEFAULT_THREADNUM),
-      _cond(threadNum ? threadNum : DEFAULT_THREADNUM),
-      _push(true),
-      _run(false),
-      _activeThreadCount(0) {
-    if (_threadNum == 0) {
-        _threadNum = DEFAULT_THREADNUM;
-    }
-}
+            if (!workerItem.fn && !_queues[id].pop(workerItem)) {
+                break;
+            }
 
-inline ThreadPool::~ThreadPool() { stop(STOP_AND_CLEAR_QUEUE); }
+            if (workerItem.fn) {
```

Comment:
We could eliminate such braces.

Reply:
If the user pass nullptr to threadpool workerItem.fn() will crash.
Do you mean use ```assert(workerItem.fn);``` instead of null checking?